### PR TITLE
fileserver: properly handle escaped/non-ascii paths

### LINKED ISSUE #4329: 404 on filenames with accents
Hi there,

Caddy gives me an error 404 for files with accents like `cropped-Capture-d’écran-2019-12-02-à-16.22.37.png` (yes, I know, this is awfull), which are called `cropped-Capture-d%E2%80%99%C3%A9cran-2019-12-02-%C3%A0-16.22.37.png` by the browser.

Obviously the file does exists and the files without accents gives a `200 OK`.

Log file is 
```json
{
    "level": "error",
    "ts": 1630683589.3759747,
    "logger": "http.log.access.log16",
    "msg": "handled request",
    "request": {
        "remote_addr": "91.180.123.81:51607",
        "proto": "HTTP/2.0",
        "method": "GET",
        "host": "[REDACTED]",
        "uri": "/wp-content/uploads/2019/12/cropped-Capture-d%E2%80%99%C3%A9cran-2019-12-02-%C3%A0-16.22.37.png",
        "headers": {
            "Upgrade-Insecure-Requests": [
                "1"
            ],
            "Sec-Fetch-Dest": [
                "document"
            ],
            "User-Agent": [
                "Mozilla/5.0 (Macintosh; Intel Mac OS X 10_15_7) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/93.0.4577.63 Safari/537.36"
            ],
            "Accept": [
                "text/html,application/xhtml+xml,application/xml;q=0.9,image/avif,image/webp,image/apng,*/*;q=0.8,application/signed-exchange;v=b3;q=0.9"
            ],
            "Sec-Fetch-User": [
                "?1"
            ],
            "Accept-Encoding": [
                "gzip, deflate, br"
            ],
            "Accept-Language": [
                "fr-FR,fr;q=0.9,en-US;q=0.8,en;q=0.7"
            ],
            "Sec-Ch-Ua": [
                "\"Google Chrome\";v=\"93\", \" Not;A Brand\";v=\"99\", \"Chromium\";v=\"93\""
            ],
            "Sec-Ch-Ua-Mobile": [
                "?0"
            ],
            "Sec-Fetch-Site": [
                "none"
            ],
            "Sec-Fetch-Mode": [
                "navigate"
            ],
            "Cookie": [
                "wp-settings-time-2=1617279544; wordpress_test_cookie=WP+Cookie+check"
            ],
            "Sec-Ch-Ua-Platform": [
                "\"macOS\""
            ],
            "Dnt": [
                "1"
            ]
        },
        "tls": {
            "resumed": true,
            "version": 772,
            "cipher_suite": 4865,
            "proto": "h2",
            "proto_mutual": true,
            "server_name": "[REDACTED]",
        }
    },
    "common_log": "91.180.123.81 - - [03/Sep/2021:17:39:49 +0200] \"GET /wp-content/uploads/2019/12/cropped-Capture-d%E2%80%99%C3%A9cran-2019-12-02-%C3%A0-16.22.37.png HTTP/2.0\" 404 12617",
    "user_id": "",
    "duration": 0.884501508,
    "size": 12617,
    "status": 404,
    "resp_headers": {
        "Expires": [
            "Wed, 11 Jan 1984 05:00:00 GMT"
        ],
        "Cache-Control": [
            "no-cache, must-revalidate, max-age=0"
        ],
        "Link": [
            "<https://[REDACTED]/wp-json/>; rel=\"https://api.w.org/\""
        ],
        "Status": [
            "404 Not Found"
        ],
        "Content-Encoding": [
            "gzip"
        ],
        "Vary": [
            "Accept-Encoding"
        ],
        "Server": [
            "Caddy"
        ],
        "Content-Type": [
            "text/html; charset=UTF-8"
        ]
    }
}
```

The config file is the following
```
www.[REDACTED].be {
    redir https://[REDACTED]{uri} permanent
}

[REDACTED].be {
    root * /var/www/redacted
    php_fastcgi unix//run/php/php7.3-fpm.sock
    file_server
    encode gzip
    @disallowed {
        path /xmlrpc.php
        path *.sql
        path /wp-content/uploads/*.php
    }
    rewrite @disallowed '/index.php'

    log {
        output file /var/log/caddy/[REDACTED].be.log
    }

    tls backend@[REDACTED] {
        protocols tls1.2 tls1.3
    }
}
```

### PR DESCRIPTION
**Pull request by Mohammed90**
I might add integration test for those non-ascii file names. Thoughts?

Closes #4329

### CODE REVIEW

**Comment by Mohammed90** [comment]
Windows fails because the colon `:` on NTFS filesystem creates data streams and not typically allowed in file names. I'll work around that.

File: modules/caddyhttp/fileserver/staticfiles.go [hunk 1]
```diff
@@ -165,6 +166,16 @@ func (fsrv *FileServer) ServeHTTP(w http.ResponseWriter, r *http.Request, next c
 	filesToHide := fsrv.transformHidePaths(repl)
 
 	root := repl.ReplaceAll(fsrv.Root, ".")
+	// PathUnescape returns an error if the escapes aren't well-formed,
+	// meaning the count % matches the RFC. Return early if the escape is
+	// improper.
+	if _, err := url.PathUnescape(r.URL.Path); err != nil {
```
Comment (francislavoie) [inline modules/caddyhttp/fileserver/staticfiles.go:172]: Should we use https://pkg.go.dev/net/url#URL.EscapedPath instead? 🤔

**Review by mholt — APPROVED** [review]
Thanks for the fix, I like the implementation, seems to do the right things at the right places, and the tests look good too.

If we're feeling good about this, feel free to merge it. :+1:

**Review by francislavoie — APPROVED** [review]
I've tried to mentally poke holes in this, but I haven't been able to. I think it should be good. Still makes me 5% nervous but it's probably unfounded. :man_shrugging: Let's merge!

**Pull request merged**